crates/bevy_ecs/src/storage/sparse_set.rs

```diff
@@ -0,0 +1,499 @@
+
+    pub(crate) fn clear_flags(&mut self) {
+        let flags = self.flags.get_mut().iter_mut();
+        for component_flags in flags {
+            *component_flags = ComponentFlags::empty();
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SparseSet<I, V: 'static> {
+    dense: Vec<V>,
+    indices: Vec<I>,
+    sparse: SparseArray<I, usize>,
+}
+
+impl<I: SparseSetIndex, V> Default for SparseSet<I, V> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+impl<I, V> SparseSet<I, V> {
+    pub const fn new() -> Self {
+        Self {
+            dense: Vec::new(),
+            indices: Vec::new(),
+            sparse: SparseArray::new(),
+        }
+    }
+}
+
+impl<I: SparseSetIndex, V> SparseSet<I, V> {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            dense: Vec::with_capacity(capacity),
+            indices: Vec::with_capacity(capacity),
+            sparse: Default::default(),
+        }
+    }
+
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.dense.capacity()
+    }
+
+    pub fn insert(&mut self, index: I, value: V) {
+        if let Some(dense_index) = self.sparse.get(index.clone()).cloned() {
+            // SAFE: dense indices stored in self.sparse always exist
+            unsafe {
+                *self.dense.get_unchecked_mut(dense_index) = value;
+            }
+        } else {
+            self.sparse.insert(index.clone(), self.dense.len());
+            self.indices.push(index);
+            self.dense.push(value);
+        }
+
+        // PERF: switch to this. it's faster but it has an invalid memory access on table_add_remove_many
+        // let dense = &mut self.dense;
+        // let indices = &mut self.indices;
+        // let dense_index = *self.sparse.get_or_insert_with(index.clone(), move || {
+        //     if dense.len() == dense.capacity() {
+        //         dense.reserve(64);
+        //         indices.reserve(64);
+        //     }
+        //     let len = dense.len();
+        //     // SAFE: we set the index immediately
+        //     unsafe {
+        //         dense.set_len(len + 1);
+        //         indices.set_len(len + 1);
+        //     }
+        //     len
+        // });
+        // // SAFE: index either already existed or was just allocated
+        // unsafe {
+        //     *self.dense.get_unchecked_mut(dense_index) = value;
+        //     *self.indices.get_unchecked_mut(dense_index) = index;
+        // }
+    }
+
+    pub fn get_or_insert_with(&mut self, index: I, func: impl FnOnce() -> V) -> &mut V {
+        if let Some(dense_index) = self.sparse.get(index.clone()).cloned() {
+            // SAFE: dense indices stored in self.sparse always exist
+            unsafe { self.dense.get_unchecked_mut(dense_index) }
+        } else {
+            let dense_index = self.dense.len();
+            self.sparse.insert(index.clone(), dense_index);
+            self.indices.push(index);
+            self.dense.push(func());
```

Comment:
What if `func` panics? The `self.dense` won't contain the value, but using `catch_unwind` it is still possible to access the sparse set. Try moving the `func()` to above the `self.spare.insert()` call.